proposals/NNNN-extension-macros.md

```diff
@@ -0,0 +1,201 @@
+
+## Proposed solution
+
+This proposal removes the `conformance` macro role in favor of an `extension` macro role. An `extension` macro role can be used with the `@attached` macro attribute, and it can add a conformance, a `where` clause, and a member list in an extension on the type the macro is attached to:
+
+```swift
+protocol MyProtocol {
+  func requirement
+}
+
+@attached(extension, conformances: MyProtocol, names: named(requirement))
+macro MyProtocol = #externalMacro(...)
+
+@MyProtocol
+struct S<T> {}
+
+// expands to
+
+extension S: MyProtocol where T: MyProtocol {
+  func requirement() { ... }
+}
+```
+
+The generated extensions of the macro must only extend the type the macro is attached to. Any conformances or members must also be specified upfront by the `@attached(extension)` attribute.
+
+## Detailed design
+
+### Specifying macro-introduced protocol conformances and member names
+
+SE-0389 states that whenever a macro produces declarations that are visible to other Swift code, it is required to declare the names in advance. This rule also applies to extension macros, which must specify:
+
+* Declarations inside the extension, which can be specified using `named`, `prefixed`, `suffixed`, and `arbitrary`.
+* The names of protocols that are listed in the extension's conformance clause. These protocols are specified in the `conformances:` list of the `@attached(conformances:)` attribute. Each name that appears in this list must be a conformance constraint, where a conformance constraint is one of:
+  * A protocol name
+  * A typealias whose underlying type is a conformance constraint
+  * A protocol composition whose entires are each a conformance constraint
+
+The following restrictions apply to generated conformances and names listed in `@attached(extension)`:
+
+* An extension macro cannot add a conformance to a protocol that is not covered by the `conformances:` list in `@attached(extension, conformnaces:)`.
+* An extension macro cannot add a member that is not covered by the `names:` list in `@attached(extension, names:)`.
+* An extension macro cannot introduce an extension with an attached `peer` macro, because the peer-macro-generated names are not covered by the original `@attached(extension)` attribute.
+
+### Extension macro application
+
+Extension macros can only be attached to the primary declaration of a nominal type; they cannot be attached to typealias or extension declarations.
+
+
+Extensions are only valid at the top-level. When an extension macro is applied to a nested type:
+
+```swift
+@attached(extension, conformances: MyProtocol, names: named(requirement))
+macro MyProtocol = #externalMacro(...)
+
+struct Outer {
+  @MyProtocol
+  struct Inner {}
+}
+```
+
+The macro expansion containing the extension is inserted at the top-level. The above code expands to:
```

Comment:
```suggestion
In this situation, the macro expansion containing the `extension` is inserted at the top level of the file, instead of immediately where the macro is invoked, where the `extension` would be invalid. The above code expands to:
```